### ISSUE TITLE
HMM Tagging is not accurate?

### ISSUE THREAD

**Issue by blumonkey** [nltk/nltk#1095]
Hello, I have been trying to implement a simple POS tagger using HMM and came up with the following code.

```
 import nltk
 from nltk.corpus import treebank

train_data = treebank.tagged_sents()[:3000]

print train_data[0]
# [(u'Pierre', u'NNP'), (u'Vinken', u'NNP'), (u',', u','), (u'61', u'CD'), (u'years', u'NNS'), (u'old', u'JJ'), (u',', u','), ... ]

from nltk.tag import hmm

trainer = hmm.HiddenMarkovModelTrainer()
tagger = trainer.train_supervised(train_data)

print tagger

print tagger.tag("Alex was born in Connecticut .".split())
# [('Alex', u'NNP'), ('was', u'NNP'), ('born', u'NNP'), ('in', u'NNP'), ('Connecticut', u'NNP'), ('.', u'NNP')]

print tagger.tag("Joe met Joanne in Delhi .".split())
# [('Joe', u'NNP'), ('met', u'VBD'), ('Joanne', u'NNP'), ('in', u'IN'), ('Delhi', u'NNP'), ('.', u'NNP')]

print tagger.tag("Chicago is the birthpace of Ginny".split())
# [('Chicago', u'NNP'), ('is', u'VBZ'), ('the', u'DT'), ('birthpace', u'NNP'), ('of', u'NNP'), ('Ginny', u'NNP')]
```

As you can see (many)tags are pretty much off. Is there something wrong in my approach or could this be a bug?

Also, there is only a 0.3 match with the gold standard when I run `tagger.evaluate(treebank.tagged_sents()[3000:])`

**Comment by alvations** [comment]
@blumonkey Thank you for your question. 

Possibly, it's more of a `train vs test` and overfitting issue. If we look at the tagger's performance on the training set we see that it's at 98%. 

```python
# Evaluating on train data
>>> tagger.evaluate(treebank.tagged_sents()[:3000])
0.981615512636916

# Evaluating on test data.
>>> tagger.evaluate(treebank.tagged_sents()[3000:])
0.36844377293330455
```

The low error rate on train and high error rate on test, suggests that:

 - The model is overfitting to the training data
 - The training data is insufficient to cover the cases in the test data
 - Using raw sentence tokens as your input to classifier is not sufficient, you might have to do normalization, lemmatization, stemming, etc. 

If we look at the out-of-vocabulary unigrams in test that are not available in the training data we see that we essentially have a high OOV of 35%.

```python
>>> train_data = list(treebank.tagged_sents()[:3000])
>>> test_data = list(treebank.tagged_sents()[3000:])
>>> train_words, train_tags = zip(*chain(*train_data))
>>> test_words, test_tags = zip(*chain(*test_data))
>>> len(set(test_words))
4555
>>> len(set(test_words).difference(train_words))
1629
```

Also, if we stem the data, we immediately see a jump to 44%

```python
>>> from nltk.stem import PorterStemmer
>>> porter = PorterStemmer()

>>> train_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in treebank.tagged_sents()[:3000]]
>>> test_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in treebank.tagged_sents()[3000:]]

>>> train_words, train_tags = zip(*chain(*train_data))
>>> test_words, test_tags = zip(*chain(*test_data))
>>> len(set(test_words).difference(train_words))
1121

>>> trainer = hmm.HiddenMarkovModelTrainer()
>>> tagger = trainer.train_supervised(train_data)
>>> tagger.evaluate(test_data)
0.44036261601554066
```

When I tried to randomize the data, I saw another jump in accuracy. And if we do more shuffling, we see that we're hitting the plateau at high 40s, low 50s but not exceeding. So the model isn't generalizing enough.

```python
>>> import random
>>> all_data = list(treebank.tagged_sents())
>>> random.shuffle(all_data)
>>> train_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[:3000]]
>>> test_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[3000:]]
>>> trainer = hmm.HiddenMarkovModelTrainer()
>>> tagger = trainer.train_supervised(train_data)
>>> tagger.evaluate(test_data)
0.531570656107634

>>> random.shuffle(all_data)
>>> train_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[:3000]]
>>> test_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[3000:]]
>>> trainer = hmm.HiddenMarkovModelTrainer()
>>> tagger = trainer.train_supervised(train_data)
>>> tagger.evaluate(test_data)
0.5417020553762527

>>> for _ in range(10):
...     random.shuffle(all_data)
...     train_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[:3000]]
...     test_data = [ [(porter.stem(word.lower()), tag) for word, tag in sent] for sent in all_data[3000:]]
...     trainer = hmm.HiddenMarkovModelTrainer()
...     tagger = trainer.train_supervised(train_data)
...     print(tagger.evaluate(train_data), tagger.evaluate(test_data))
... 
(0.9200600196616133, 0.5429647381033893)
(0.9198430927249736, 0.5018983518854085)
(0.9177755915704447, 0.5184853142076503)
(0.9198630761020942, 0.49804990564059554)
(0.9199244201198853, 0.5116978609625669)
(0.919353146394565, 0.506713057715923)
(0.9189899984453542, 0.5248211852861036)
(0.9197032258064516, 0.5377114256127028)
```

Note that the `treebank` corpus in NLTK is a sample of the full Penn Treebank, please get the full access of the PTB from https://catalog.ldc.upenn.edu/ldc99t42

To improve the model, try retraining with the full Penn Treebank.

If more data is still failing on the Penn Treebank, possibly you will have to look into why the model is failing in more detail. 

----

**Comment by Ligia-M** [comment]
thank you @alvations, just tried your suggestions and my test accuracy went from .72 to .92!!!!!! 

Could you please explain to me why randomizing the data increases the accuracy? (does it indirectly increase the generalizability of the model?) 
Thank you and have a good day.
